syntax = "proto3";

package bookverse.books;

import "google/protobuf/empty.proto";
import "google/protobuf/timestamp.proto";
import "google/protobuf/field_mask.proto";

message Book {
  int32 id = 1;
  string title = 2;
  string author = 3;
  string isbn = 4;
  int32 published_year = 5;
  string genre = 6;
  string description = 7;
  google.protobuf.Timestamp created_at = 8;
  google.protobuf.Timestamp updated_at = 9;

}

message BookId {
  int32 id = 1;
}

message CreateBookRequest {
  string title = 1;
  string author = 2;
  string isbn = 3;
  int32 published_year = 4;
  string genre = 5;
  string description = 6;
}

message CreateBookResponse {
  Book book = 1;
  bool success = 2;
  string message = 3;
}

message GetBookResponse {
  Book book = 1;
  bool found = 2;
}

message UpdateBookRequest {
  int32 id = 1;
  Book book = 2;  // The updated book data
  google.protobuf.FieldMask update_mask = 3;  // Which fields to update
}

message UpdateBookResponse {
  Book book = 1;
  bool success = 2;
  string message = 3;
}

message DeleteBookResponse {
  bool success = 1;
  string message = 2;
}

message ListBooksRequest {
  int32 limit = 1;
  int32 offset = 2;
  string genre_filter = 3;
  string author_filter = 4;
}

message ListBooksResponse {
  repeated Book books = 1;
  int32 total_count = 2;
}


service BooksService{
  rpc CreateBook (CreateBookRequest) returns (CreateBookResponse);
  rpc GetBook (BookId) returns (GetBookResponse);
  rpc UpdateBook (UpdateBookRequest) returns (UpdateBookResponse);
  rpc DeleteBook (BookId) returns (DeleteBookResponse);
  rpc ListBooks (ListBooksRequest) returns (ListBooksResponse);
  rpc GetAllBooks (google.protobuf.Empty) returns (ListBooksResponse);
}